### ISSUE TITLE
Weird:   Submitted single threaded is about as slow as multithreaded

### ISSUE THREAD

**Issue by jimduff-usds** [CDCgov/prime-reportstream#1608]
@SpiderSense did some testing yesterday, wherein a single thread submitting reports as fast as it can, was able to submit about **80 or so reports in a minute.**

And 8 threads, again running as fast as they can, are able to submit **110 reports per minute.**

Heh??? 8 threads is only 25% faster than one?

Steps:

FIrst, write a test that let's us measure this problem:

- Write a new Parallel() test in TestReportStream that takes a --threads <T> argument, and runs that many threads, each submitting --submits <S> reports, each with --items <I> items.
- note the --submits and --items args already exist, --threads is new.
- Unlike all other tests, generate fake data just once, and use it over and over, so that fake data generation time doesn't slow it down.
- Each thread should wait for the response, check that its a 201, then submit again.  No sleep in between.
- Carefully time things, and print out useful timing info, as total number of submits per minute, across all threads.  eg, 
"With T threads running, ReportStream was able to submit reports with I items, at a pace of R reports per minute.  Each thread submitted S reports".

Then, figure out what's going on, and propose fixes.

**Comment by td-usds** [comment]
Adding myself @td-usds

**Comment by td-usds** [comment]
Current findings:
We looked at the logs and think that it is indeed a timeout caused by trying to get a connection to the DB which is configured in code (and so not likely an infrastructure bug)
if you look at `prime-router/src/main/kotlin/azure/DatabaseAccess.kt` line 667 and line 675

* 667 should probably be `config.setConnectionTimeout(60000)`
* 675 should probably be increased significantly to be however many concurrent requests are handled by a single instance of the function app (give or take some overhead); this fix will probably also include changing the `prime-router/host.json` file to explicitly specify how many concurrent requests are dealt with by a single instance

And then on the infrastructure side, we should/will also add an autoscaling rule that trips on concurrent request count (currently we only have CPU and memory pressure)

**Comment by td-usds** [comment]
Pablo is going to: 1) re-run Giang’s test table, but on Pablo’s own laptop. 2) Study the latencies for the very most simple case: 1 thread, submitting 1 row. 3) Study the case of, say, 4 threads — now where are the latencies?
For Thomas: 1) If I hit the system with 8 client requests, does it start 8 ReportFunctions? 2) How many active database connections are in use when 1,2,4,8 threads are hitting the system.

**Comment by td-usds** [comment]
> 1) If I hit the system with 8 client requests, does it start 8 ReportFunctions?

Yes and no: yes it will handle the 8 requests in parallel but there is no guarantee that this is by 8 different instances of the container running our code. In fact, it may handle all 8 in parallel in a single container

> 2) How many active database connections are in use when 1,2,4,8 threads are hitting the system.

TBD

**Comment by jimduff-usds** [comment]
@carlosfelix2  - a couple important notes:

Running with just one thread, in Staging, looking at the raw logs, the main slowness is between the line
`Successfully reported: <reportid>.`
and the line
`Queue: receiver&BATCH&waters.giang&2021-08-07T08:49:00-04:00`

See attached screenshot showing two submissions - they all seem to fit this same pattern of around 1/2 second delay between those two logging statements.

#### This screenshot shows two submissions from Giang's test to Staging, submitting from just a single thread.  Note the long delay (about 1/2 sec) between those two statements mentioned above:
<img width="1323" alt="image" src="https://user-images.githubusercontent.com/62258514/128602088-eeced89a-31f9-45f5-a984-edae1caf0b7d.png">

So, that's a good first clue.

Second, I noted that Giang was using two test receivers in Staging that are not in organizations.yml, which is part of the reason why localhost is faster than Staging.    If you approve and merge my PR #1807 , you'll see I have conveniently added in giant's test receivers to organizations.yml.   Once you add those, his tests do run slower, but still not quite as slow as Staging.   But **the really good news is**, even though its less pronounced, I'm seeing the _same delay locally between those two logging statements_, so you can do profiling locally without having to futz around with Staging.

All the above analysis is with just a single thread submitting.

**Comment by jimduff-usds** [comment]
Now some results from running 4 threads in staging:

I ran `jmeter -n -t DataRobot_ReportStream_Load.jmx -Jthreads=4 -Jprotocol=https -Jserver=staging.prime.cdc.gov -Jport=443 -l results.jtl `

Here is a screenshot of the beginning of the run:
<img width="1263" alt="image" src="https://user-images.githubusercontent.com/62258514/128606247-cf631fe2-85c7-4ea9-b266-810ad37b8f2d.png">

If you look at this screenshot, from the very beginning of the run, you'll see all 4 threads starting up in parallel - lovely!  However, if you look at lines 13-14, you'll see that the delay between 
`Successfully reported: 2dc750bc-33c2-4e2c-bfc3-879a57985c11.`
and
`Queue: receiver&BATCH&waters.giang&2021-08-07T11:33:00-04:00`
_is almost two seconds -_ that is, whatever is occurring between those steps is happening serially.   One would guess the transaction on `ReportFunction.kt line 397` is a likely culprit.

**Comment by jimduff-usds** [comment]
A bit more investigation:

I did a little bit more looking.  Surprisingly it appears that the slowness mentioned above is primarily in-memory processing of all rows times all receivers to do all the filtering and transforms, in function filterAndTranslateByReceiver.

I'll be interested in whether you see the same thing.

But its perfectly fine if in-memory processing is slow, as long as we can operate in parallel, which it appears we cannot.   So the problem is that the slow stuff is happening inside something that's forcing it to run serially.   Again, it seems likely the database transaction is the culprit.

**Comment by RickHawesUSDS** [comment]
Perhaps, it is time to remove the outer transaction? I took some care to ensure that it wasn't table locking when I wrote it, but really did it in a good enough fashion.

**Comment by jimduff-usds** [comment]
After @carlosfelix2 's PR #1849, on my local Mac, I got 325 submissions in 2 minutes with 1 thread and 635 submissions in 2minutes with 4 threads.  That's a definite improvement.
